File: 345.reverse-vowels-of-a-string.java
/*
 * @lc app=leetcode id=345 lang=java
 *
 * [345] Reverse Vowels of a String
 */
class Solution {
    public String reverseVowels(String s) {
        if (s == null || s.length() == 0) return s;
        char[] ss = s.toCharArray();
        int i = 0, j = s.length() - 1;
        while (i < j) {
            while (i < j && !isVowel(ss[i])) i++;
            while (i < j && !isVowel(ss[j])) j--;
            if (i < j) {
                char tmp = ss[i];
                ss[i++] = ss[j];
                ss[j--] = tmp;
            }
        }
        return String.valueOf(ss);
    }

    private boolean isVowel(char c) {
        switch (c) {
            case 'a':
            case 'e':
            case 'i':
            case 'o':
            case 'u':
            case 'A':
            case 'E':
            case 'I':
            case 'O':
            case 'U':
                return true;
            default:
                return false;
        }
    }
}

File: go/345.reverse-vowels-of-a-string.go
import "strings"

/*
 * @lc app=leetcode id=345 lang=golang
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start

const vowels = "aeiouAEIOU"

func reverseVowels(s string) string {
	i, j := 0, len(s)-1
	rs := []rune(s)
OUTER:
	for i < j {
		for !strings.Contains(vowels, string(rs[i])) {
			i++
			if i >= j {
				break OUTER
			}
		}
		for !strings.Contains(vowels, string(rs[j])) {
			j--
			if i >= j {
				break OUTER
			}
		}
		rs[i], rs[j] = rs[j], rs[i]
		i++
		j--
	}
	return string(rs)
}

// @lc code=end

